zrml/prediction-markets/src/benchmarks.rs

```diff
@@ -169,18 +169,22 @@ fn setup_redeem_shares_common<T: Config + pallet_timestamp::Config>(
     };
     let grace_period: u32 =
         (market.deadlines.grace_period.saturated_into::<u32>() + 1) * MILLISECS_PER_BLOCK;
-    pallet_timestamp::Pallet::<T>::set_timestamp((end + grace_period).into());
+    let block = frame_system::Pallet::<T>::block_number();
+    zeitgeist_utils::set_block_number_timestamp::<T>(block, (end + grace_period).into());
     Call::<T>::report { market_id, outcome }
         .dispatch_bypass_filter(RawOrigin::Signed(caller.clone()).into())?;
     Call::<T>::admin_move_market_to_resolved { market_id }
         .dispatch_bypass_filter(resolve_origin)?;
     Ok((caller, market_id))
 }
 
```

Comment:
Thanks for applying the patch at https://github.com/zeitgeistpm/zeitgeist/pull/887#discussion_r1072825071 . I see now what was missing in the patch to make it complete. Some arguments against the approach to tightly couple Aura:

- The problem actually lies in the minimum timestamp period. It is too low in the benchmark environment, which hinders us from "skipping" blocks in terms of a timestamp delta. This is does not only affect Aura, it could also deny proper execution of other consensus systems.
- Currently we enforce that projects set the `parachain` feature when using PM in case they don't use Aura.
- Projects that have a standalone chain but don't use Aura (they could use Babe instead for example), won't be able to use the pallet benchmarks.

The solution suggested here does two things:
1. Create a custom `MinimumPeriod` implementation.
2. Within the `MinimumPeriod` implementation, figure out whether the node is in a benchmark environment or not and return an adequate value.

Unfortunately I am not aware of a direct way to query whether the node is in a benchmark environment, however after some extensive analysis of the pallet there is an indirect way. The [account whitelist can be fetched](https://github.com/paritytech/substrate/blob/d4384f242f48fd62516aac02c3f18b5a8c9e5793/frame/benchmarking/src/utils.rs#L281-L284), and this is something that [should only be set in a benchmark environment](https://github.com/paritytech/substrate/blob/c8653447fc8ef8d95a92fe164c96dffb37919e85/primitives/externalities/src/lib.rs#L289-L294). During initialization of the benchmarks, a couple of keys are added to the whitelist. We can use this as a measurement to figure out whether the node is in a benchmark environment, by checking whether the whitelist contains elements.

The solution is to define the `MinimumPeriod` of the timestamp pallet as such:
```rust
// Timestamp
/// Custom getter for minimum timestamp delta.
/// This ensures that consensus systems like Aura don't experience breaking assertions
/// in a benchmark environment
pub struct MinimumPeriod;
impl MinimumPeriod {
    /// Returns the value of this parameter type.
    pub fn get() -> u64 {
        #[cfg(feature = "runtime-benchmarks")]
        {
            use frame_benchmarking::{Benchmarking, benchmarking::get_whitelist};
            // Should that condition be true, we can assume that we are in a benchmark environment.
            if !get_whitelist().is_empty() { return u64::MAX; }
        }


        MILLISECS_PER_BLOCK as u64 / 2
    }
}
impl<I: From<u64>> frame_support::traits::Get<I> for MinimumPeriod {
    fn get() -> I {
        I::from(Self::get())
    }
}
impl frame_support::traits::TypedGet for MinimumPeriod {
    type Type = u64;
    fn get() -> u64 {
        Self::get()
    }
}
```

You can add https://github.com/sea212/zeitgeist as a remote and then merge the branch `sea212-fix-aura-slot-issue` into this PR. [Click here to see a changelog](https://github.com/zeitgeistpm/zeitgeist/compare/zeitgeistpm:zeitgeist:upgrade_0929...sea212:zeitgeist:sea212-fix-aura-slot-issue).

Reply:
I am not sure if the changes (besides the Aura thing) within the PM benchmarks were mandatory, so feel free to revert some of the changes from my PR in regards to that.

Reply:
It feels like with this new solution it will always have CurrentSlot as zero. and that's how it prevents assertion failures.

Reply:
I had initially though for making something like that but when I took look at other projects their approach seems better to me. :)